### PR TITLE
Improve UX if no speech to text app is installed

### LINKED ISSUE #1323: After installing voice recognition app, widgets still don't work
### Actual behaviour
If no voice recognizer is installed at widget creation time, a click on the mic opens the play store with the Google app entry. After installing this app, the mic still opens the play store.

### Expected behaviour
The UX should be improved, e.g. by blocking the widget creation if no voice recognition app is installed. Open Play Store or a dialog instead.

### PR DESCRIPTION
**Pull request by mueller-ma**
* Hide voice widgets
* Always show mic in action bar, but open snackbar with link to Google's
stt app if no stt app is installed

Fixes #1323

Signed-off-by: mueller-ma <mueller-ma@users.noreply.github.com>

### CODE REVIEW

File: mobile/src/main/java/org/openhab/habdroid/ui/MainActivity.java [hunk 1]
```diff
@@ -858,13 +861,28 @@ public boolean onPrepareOptionsMenu(Menu menu) {
...
-        manageVoiceRecognitionShortcut(SpeechRecognizer.isRecognitionAvailable(this));
+        voiceRecognitionItem.setVisible(mConnection != null);
         voiceRecognitionItem.getIcon().setColorFilter(iconColor, PorterDuff.Mode.SRC_IN);
+
+        boolean isSpeechRecognizerAvailable = SpeechRecognizer.isRecognitionAvailable(this);
+        manageVoiceRecognitionShortcut(isSpeechRecognizerAvailable);
+        setVoiceWidgetComponentEnabledSetting(VoiceWidget.class.getCanonicalName(),
+                isSpeechRecognizerAvailable);
+        setVoiceWidgetComponentEnabledSetting(VoiceWidgetWithIcon.class.getCanonicalName(),
+                isSpeechRecognizerAvailable);
```
Comment (maniac103) [inline mobile/src/main/java/org/openhab/habdroid/ui/MainActivity.java]: Doing this every time when populating the menu seems a bit excessive. I'd suggest doing this in an AsyncTask fired from onCreate() instead.

File: mobile/src/main/java/org/openhab/habdroid/ui/MainActivity.java [hunk 2]
```diff
@@ -858,13 +861,28 @@ public boolean onPrepareOptionsMenu(Menu menu) {
...
         return true;
     }
 
+    private void setVoiceWidgetComponentEnabledSetting(String className,
+            boolean isSpeechRecognizerAvailable) {
+        ComponentName voiceWidget = new ComponentName(BuildConfig.APPLICATION_ID, className);
+        PackageManager pm  = getApplicationContext().getPackageManager();
```
Comment (maniac103) [inline mobile/src/main/java/org/openhab/habdroid/ui/MainActivity.java]: Excessive space.

File: mobile/src/main/java/org/openhab/habdroid/ui/MainActivity.java [hunk 3]
```diff
@@ -858,13 +861,28 @@ public boolean onPrepareOptionsMenu(Menu menu) {
...
+        boolean isSpeechRecognizerAvailable = SpeechRecognizer.isRecognitionAvailable(this);
+        manageVoiceRecognitionShortcut(isSpeechRecognizerAvailable);
+        setVoiceWidgetComponentEnabledSetting(VoiceWidget.class.getCanonicalName(),
+                isSpeechRecognizerAvailable);
+        setVoiceWidgetComponentEnabledSetting(VoiceWidgetWithIcon.class.getCanonicalName(),
+                isSpeechRecognizerAvailable);
         return true;
     }
 
+    private void setVoiceWidgetComponentEnabledSetting(String className,
```
Comment (maniac103) [inline mobile/src/main/java/org/openhab/habdroid/ui/MainActivity.java]: Maybe pass the class as `Class<?>` and convert to string here?

File: mobile/src/main/java/org/openhab/habdroid/ui/MainActivity.java [hunk 4]
```diff
@@ -858,13 +861,28 @@ public boolean onPrepareOptionsMenu(Menu menu) {
...
+            boolean isSpeechRecognizerAvailable) {
+        ComponentName voiceWidget = new ComponentName(BuildConfig.APPLICATION_ID, className);
+        PackageManager pm  = getApplicationContext().getPackageManager();
+        pm.setComponentEnabledSetting(voiceWidget,
+                isSpeechRecognizerAvailable ? PackageManager.COMPONENT_ENABLED_STATE_ENABLED
+                        : PackageManager.COMPONENT_ENABLED_STATE_DISABLED,
```
Comment (maniac103) [inline mobile/src/main/java/org/openhab/habdroid/ui/MainActivity.java]: I think having an intermediate variable would make sense here.

File: mobile/src/main/java/org/openhab/habdroid/ui/MainActivity.java [hunk 5]
```diff
@@ -994,13 +1012,24 @@ private void launchVoiceRecognition() {
...
             startActivity(speechIntent);
         } catch (ActivityNotFoundException speechRecognizerNotFoundException) {
-            try {
-                startActivity(new Intent(Intent.ACTION_VIEW,
-                        Uri.parse("market://details?id=com.google.android.googlequicksearchbox")));
-            } catch (ActivityNotFoundException appStoreNotFoundException) {
-                Toasty.error(this, R.string.error_no_app_store_found,
-                        Toast.LENGTH_LONG, true).show();
-            }
+                mLastSnackbar = Snackbar.make(findViewById(android.R.id.content),
```
Comment (maniac103) [inline mobile/src/main/java/org/openhab/habdroid/ui/MainActivity.java]: Shouldn't you close previous snackbar (if present) here ...

File: mobile/src/main/java/org/openhab/habdroid/ui/MainActivity.java [hunk 6]
```diff
@@ -994,13 +1012,24 @@ private void launchVoiceRecognition() {
...
+                mLastSnackbar.setAction(R.string.install, v -> {
+                    try {
+                        startActivity(new Intent(Intent.ACTION_VIEW,
+                                Uri.parse("market://details?id=com.google.android.googlequicksearchbox")));
+                    } catch (ActivityNotFoundException appStoreNotFoundException) {
+                        try {
+                            startActivity(new Intent(Intent.ACTION_VIEW, Uri.parse(
+                                    "http://play.google.com/store/apps/details?id=com.google.android.googlequicksearchbox")));
+                        } catch (ActivityNotFoundException browserNotFoundException) {
+                            mLastSnackbar = Snackbar.make(findViewById(android.R.id.content),
```
Comment (maniac103) [inline mobile/src/main/java/org/openhab/habdroid/ui/MainActivity.java]: ... and here?

File: mobile/src/main/java/org/openhab/habdroid/ui/MainActivity.java (hunk 5 above)
Comment (mueller-ma) [reply mobile/src/main/java/org/openhab/habdroid/ui/MainActivity.java]: Done

File: mobile/src/main/java/org/openhab/habdroid/ui/MainActivity.java (hunk 6 above)
Comment (mueller-ma) [reply mobile/src/main/java/org/openhab/habdroid/ui/MainActivity.java]: Done

File: mobile/src/main/java/org/openhab/habdroid/ui/MainActivity.java (hunk 4 above)
Comment (mueller-ma) [reply mobile/src/main/java/org/openhab/habdroid/ui/MainActivity.java]: Done

File: mobile/src/main/java/org/openhab/habdroid/ui/MainActivity.java (hunk 3 above)
Comment (mueller-ma) [reply mobile/src/main/java/org/openhab/habdroid/ui/MainActivity.java]: Done

File: mobile/src/main/java/org/openhab/habdroid/ui/MainActivity.java (hunk 2 above)
Comment (mueller-ma) [reply mobile/src/main/java/org/openhab/habdroid/ui/MainActivity.java]: Done

File: mobile/src/main/java/org/openhab/habdroid/ui/MainActivity.java (hunk 1 above)
Comment (mueller-ma) [reply mobile/src/main/java/org/openhab/habdroid/ui/MainActivity.java]: Done

File: mobile/src/main/java/org/openhab/habdroid/ui/MainActivity.java [hunk 7]
```diff
@@ -1010,31 +1031,34 @@ public void showRefreshHintSnackbarIfNeeded() {
...
 
-        mLastSnackbar = Snackbar.make(findViewById(android.R.id.content),
-                R.string.swipe_to_refresh_description, Snackbar.LENGTH_LONG);
-        mLastSnackbar.setAction(R.string.swipe_to_refresh_dismiss, v -> {
+        showSnackbar(R.string.swipe_to_refresh_description, R.string.swipe_to_refresh_dismiss, v -> {
             prefs.edit()
-                    .putBoolean(Constants.PREFERENCE_SWIPE_REFRESH_EXPLAINED, true)
-                    .apply();
+                .putBoolean(Constants.PREFERENCE_SWIPE_REFRESH_EXPLAINED, true)
+                .apply();
```
Comment (maniac103) [inline mobile/src/main/java/org/openhab/habdroid/ui/MainActivity.java]: Indention change looks off (we indent line continuations by 8 chars)

File: mobile/src/main/java/org/openhab/habdroid/ui/MainActivity.java [hunk 8]
```diff
@@ -1010,31 +1031,34 @@ public void showRefreshHintSnackbarIfNeeded() {
...
         final SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences(this);
-        mLastSnackbar = Snackbar.make(findViewById(android.R.id.content),
-                R.string.info_demo_mode_short, Snackbar.LENGTH_LONG);
-        mLastSnackbar.setAction(R.string.turn_off, v -> {
+        showSnackbar(R.string.info_demo_mode_short, R.string.turn_off, v -> {
             prefs.edit()
-                    .putBoolean(Constants.PREFERENCE_DEMOMODE, false)
-                    .apply();
+                .putBoolean(Constants.PREFERENCE_DEMOMODE, false)
+                .apply();
```
Comment (maniac103) [inline mobile/src/main/java/org/openhab/habdroid/ui/MainActivity.java]: As above.

File: mobile/src/main/java/org/openhab/habdroid/ui/MainActivity.java [hunk 9]
```diff
@@ -1164,4 +1188,14 @@ private void manageShortcut(boolean visible, String id, String action,
                     getString(disableMessage));
         }
     }
+
+    private void setVoiceWidgetComponentEnabledSetting(Class<?> component,
+                                                       boolean isSpeechRecognizerAvailable) {
+        ComponentName voiceWidget = new ComponentName(BuildConfig.APPLICATION_ID,
```
Comment (maniac103) [inline mobile/src/main/java/org/openhab/habdroid/ui/MainActivity.java]: Can simplify this to `new ComponentName(this, component);`

File: mobile/src/main/java/org/openhab/habdroid/ui/MainActivity.java [hunk 10]
```diff
@@ -1164,4 +1188,14 @@ private void manageShortcut(boolean visible, String id, String action,
...
+                component.getCanonicalName());
+        PackageManager pm = getApplicationContext().getPackageManager();
```
Comment (maniac103) [inline mobile/src/main/java/org/openhab/habdroid/ui/MainActivity.java]: Can simplify to `getPackageManager()` (routing via application context seems unneeded)

**Comment by mueller-ma** [comment]
Thanks, should be RTM now.

**Pull request merged**